AK/BitStream.h

```diff
@@ -150,8 +150,14 @@ class LittleEndianBitStream : public Stream {
 /// in little-endian order from another stream.
 class LittleEndianInputBitStream : public LittleEndianBitStream {
 public:
-    explicit LittleEndianInputBitStream(MaybeOwned<Stream> stream)
+    enum UnsatisfiableReadBehavior {
+        Reject,
+        FillWithZero,
+    };
+
+    explicit LittleEndianInputBitStream(MaybeOwned<Stream> stream, UnsatisfiableReadBehavior unsatisfiable_read_behavior = UnsatisfiableReadBehavior::FillWithZero)
```

Comment:
nitpick: Might be nicer to make this a template parameter to eliminate the branch when it's not needed?

Reply:
I considered it (6 months ago), but making the zeroing Stream a separate type will keep us from layering the `CanonicalCode` for `WebP` like we do at the moment (since it is shared between WebP and DEFLATE).

I don't expect it to make a big difference, if any at all. If it does, I'd rather just slap an `[[unlikely]]` on it, like we do in other places in the system in error paths.

Reply:
FWIW this PR does consistently slow down `gunzip` of https://www.mattmahoney.net/dc/enwik8.zip from 0.94s to 1.03s on my machine, about 9.5% slower.

Reply:
For me it's the following, on `/tmp` however.

Before:
```
Command:         time -n 100 unzip /home/anon/enwik8.zip
Average time:    1365.53 ms (median: 1362, stddev: 72.56, min: 1269, max:1938)
Excluding first: 1359.75 ms (median: 1360, stddev: 44.44, min: 1269, max:1529)
```

After (~1.5% slower):
```
Command:         time -n 100 unzip /home/anon/enwik8.zip
Average time:    1379.97 ms (median: 1380, stddev: 63.39, min: 1266, max:1551)
Excluding first: 1380.76 ms (median: 1380, stddev: 63.22, min: 1266, max:1551)
```

After with `[[likely]]` (~0.6% slower):
```
Command:         time -n 100 unzip /home/anon/enwik8.zip
Average time:    1373.48 ms (median: 1357.5, stddev: 76.58, min: 1290, max:2023)
Excluding first: 1366.92 ms (median: 1357, stddev: 40.25, min: 1290, max:1465)
```

Could you retest on your setup using `/tmp`, just to avoid our disk driver having any influence?

Reply:
>avoid our disk driver having any influence
Oh, my numbers are from my host, not Serenity

Reply:
Ah, Lagom doesn't have an `unzip` target, so I assumed you were talking about inside Serenity.

Reply:
For completeness, here is an implementation with template parameters (which shouldn't have any performance impact), please don't mind my naming skills :sweat_smile: :
https://github.com/SerenityOS/serenity/compare/master...LucasChollet:serenity:le_bitstream

Reply:
> Ah, Lagom doesn't have an `unzip` target, so I assumed you were talking about inside Serenity.

I'm using `gunzip`

Reply:
But `gunzip` doesn't support ZIP files, only '.gz'.

In any case, I retried with a file that I recompressed into `gzip` and can somewhat reproduce the performance loss (~5%). `[[likely]]` doesn't seem to make a significant difference either, so I guess I'll look into templates next.

Reply:
Well yeah, and I did exactly what you did - the file is just hosted online as a .zip.

Reply:
> For completeness, here is an implementation with template parameters (which shouldn't have any performance impact), please don't mind my naming skills 😅 : [master...LucasChollet:serenity:le_bitstream](https://github.com/SerenityOS/serenity/compare/master...LucasChollet:serenity:le_bitstream)

I'm not sure if I'm 100% on-board with all the `Impl` stuff, but I currently don't have enough time to try and make it look nicer either. One thing I _will_ try though is whether moving the loop into `refill_buffer_from_stream` like you did will have a positive impact on performance.

Reply:
> For completeness, here is an implementation with template parameters (which shouldn't have any performance impact), please don't mind my naming skills 😅 : [master...LucasChollet:serenity:le_bitstream](https://github.com/SerenityOS/serenity/compare/master...LucasChollet:serenity:le_bitstream)

We (apparently) decided to go with the non-templated variant for now, since moving the loop and marking the branches seems to mitigate most of the performance impact (I asked the other Tim to rerun the test, and it looks to be essentially no impact now). I added you as a coauthor on the last commit, since at least the loop moving has been taken from there.

In case this turns out to not be sufficient, we can always bring back your branch and rebase it on top of these changes.